### PR TITLE
Download as CBZ

### LINKED ISSUE #33: [Feature Request] Save chapter as CBZ
## What feature should be added to Tachidesk?
Explain What the feature is and how it should work in detail

Could you add an option save chapter as cbz ?

##  Why/Project's Benefit/Existing Problem
Explain why this should be added

Saving as CBZ will save place when downloading chaters

### PR DESCRIPTION
**Pull request by akabhirav**
Closes #33

This PR enables downloading files as CBZ. This means that Chapters will be saved as CBZ files instead of images inside the chapter folder
This PR also fixes page index issue for certain providers

### CODE REVIEW

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (akabhirav) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Some extensions do not start the index from 0. This causes the viewer and the API  to break. What I have done here is get the pages sorted by index, and then take the row at the index needed. This ensures that we never fetch a page out of order and also does not rely on the hard value of index for which tachiyomi-extension does not have any standards

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt
Comment (akabhirav) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt]: making the update page list independent of downloads makes it easier for you to restore downloads from other Tachidesk/Tachiyomi installations.

This may lead to other issues like if the server is disconnected from internet this could result in HTTP exceptions, which the downloader retry should handle properly.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]:
```suggestion
        ZipInputStream(File(cbzPath).inputStream()).use { zipInputStream ->
            var zipEntry = zipInputStream.nextEntry

            while (zipEntry != null) {
                val file = File(chapterFolder, zipEntry.name)
                if (!file.exists()) {
                    file.parentFile.mkdirs()
                    file.createNewFile()
                }

                file.outputStream().use { outputStream ->
                    zipInputStream.copyTo(outputStream)
                }
                zipEntry = zipInputStream.nextEntry
            }
        }
```

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt]: No star imports
See https://pinterest.github.io/ktlint/rules/configuration-intellij-idea/ for configuring this for your IDE

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt]: No star imports

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt]: No star imports

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt]: No star imports

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: Make sure this gets closed, with try {} finally {}

**Review by Syer10 — CHANGES_REQUESTED** [review]

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (AriaMoradi) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: The name should reflect this is a cbz chapter.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (AriaMoradi) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: This will create confusion to external users, rename to reflect that it's checking for cbz.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Put this in another PR.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt
Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt]: Put this in another PR.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt
Comment (AriaMoradi) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt]: Add a `getChapterCbzPath` and rename `getChapterDir` to `getChapterDirPath`.

use `getChapterCbzPath` instead of calculating `$chapterDir.cbz` locally throughout.

File: server/src/test/resources/server-reference.conf
Comment (AriaMoradi) [inline server/src/test/resources/server-reference.conf]: extra new line

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (AriaMoradi) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: We have a huge problem here, online image cache and download directory are the same place, it was kind of okay to ignore it up until now because downloads and cache were basically the same thing. But we can't ignore it now that downloads are not strictly in the same format as image cache.

Also there are two additional problems here.
1- The code is making the directories so if the chapter is downloaded as cbz, the next time an image is fetched it will make the directory.
2- If a user views a chapter online and then downloads some of them(az cbz or not) they would have unexpected and possibly incomplete(partially viewed) downloads.

As a result before merging this we need a PR that separates downloads from image cache(which would go into OS temp dir)

Also we need a mechanism to verify downloads exist(not deleted for example) so we won't run into this mess of a logic.

**Review by AriaMoradi — CHANGES_REQUESTED** [review]

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: or with kotlin use syntax preferably

**Comment by AriaMoradi** [comment]
I'll consider this PR blocked until we have fixed the issues with separating cache and download dir.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: Would something like `ChapterCbzFile` do?

Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: Renamed object to `ChapterCbz`

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: renamed this method to `exists`

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: What should I specifically do if closeEntry fails? Propagate the error so the downloader dies? or do something else?

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Point 2 is handled somewhat `suwayomi/tachidesk/manga/impl/ChapterCbz.kt:27`.

Regarding the separation of cache and downloads is there an upcoming PR for this, or should I try to do that?

Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Try it.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Done. https://github.com/Suwayomi/Tachidesk-Server/pull/491/files

Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Remove the change from this PR.

Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Done

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt]: The functionality is only useful with cbz downloads, as this already works with normal file downloads. In other words, this code change would be useless if cbz is not in play

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt]: Done

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/Downloader.kt]: Sorry, expanded the stars everywhere

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Chapter.kt]: Done

File: server/src/test/resources/server-reference.conf
Comment (akabhirav) [reply server/src/test/resources/server-reference.conf]: Removed

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (Syer10) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: Propagate the error, don't try to catch it, just put .closeEntry in the finally block, try catch finally doesn't require a catch.

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt
Comment (AriaMoradi) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/chapter/ChapterForDownload.kt]: alright

**Comment by AriaMoradi** [comment]
why did you close this?

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/ArchiveProvider.kt
Comment (Syer10) [inline server/src/main/kotlin/suwayomi/tachidesk/manga/impl/download/ArchiveProvider.kt]:
```suggestion
                    val entry = ZipEntry(it.name)
                    try {
                        zipOut.putNextEntry(entry)
                        it.inputStream().use { inputStream ->
                            inputStream.copyTo(zipOut)
                        }
                    } finally {
                        zipOut.closeEntry()
                    }
```

**Review by Syer10 — CHANGES_REQUESTED** [review]

**Review by Syer10 — APPROVED** [review]

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/Page.kt]: Marking this as resolved since we have already merged the changes in another PR

File: server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt
Comment (akabhirav) [reply server/src/main/kotlin/suwayomi/tachidesk/manga/impl/ChapterFile.kt]: Did this under the latest comment, sorry missed this comment. I am marking this as resolved

**Comment by kaiwei** [comment]
Thanks for this and quite excited about it. I've upgraded to 0.700 and am using the bundled WebUI but for the life of me, I can't figure out how to use the "Download as CBZ" functionality. Any hints?

**Comment by Pikcee** [comment]
> Thanks for this and quite excited about it. I've upgraded to 0.700 and am using the bundled WebUI but for the life of me, I can't figure out how to use the "Download as CBZ" functionality. Any hints?

in server.conf
```
# downloader
server.downloadAsCbz = true
```

**Comment by doctorniz** [comment]
I am getting an error when this setting is activated whilst trying to download comics

**Comment by Pikcee** [comment]
> I am getting an error when this setting is activated whilst trying to download comics

I can confirm the same here as well, please open a new issue.

**Pull request merged**
